File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldData.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldData.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

import java.util.Map;
import java.util.HashMap;

public class HistoricElementFieldData extends Element
{
	private final Map<String, Element> _fields;
	
	HistoricElementFieldData(Datetime date, Map<String, Object> values) throws Exception
	{
		this._fields = new HashMap<String, Element>();
		
		Element elmDate = new HistoricElementDateTime(date);
		this._fields.put(elmDate.name().toString(), elmDate);
		
		for (Map.Entry<String, Object> item : values.entrySet())
		{
			if(item.getValue().getClass() == Double.class)
			{
				Element elmDouble = new HistoricElementDouble(item.getKey(), (Double)item.getValue());
                this._fields.put(elmDouble.name().toString(), elmDouble);
			}
		}
	}
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("HistoricalDataRow"));
    }
	
    public Name name() throws Exception
    {
    	return new Name("fieldData");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return this._fields.size();
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
	
	public int getElementAsInt32(String name) throws Exception
	{
		return this._fields.get(name).getValueAsInt32();
	}
    
    public Datetime getElementAsDatetime(String name) throws Exception
    {
    	return this._fields.get(name).getValueAsDatetime();
    }
    
    public Datetime getElementAsDate(String name) throws Exception
    {
    	return this._fields.get(name).getValueAsDate();
    }
    
    public Datetime getElementAsTime(String name) throws Exception
    {
    	return this._fields.get(name).getValueAsTime();
    }
    
    public String getElementAsString(String name) throws Exception
    {
    	return this._fields.get(name).getValueAsString();
    }
	
	public float getElementAsFloat32(String name) throws Exception
	{
		return this._fields.get(name).getValueAsFloat32();
	}
	
	public double getElementAsFloat64(String name) throws Exception
	{
		return this._fields.get(name).getValueAsFloat64();
	}
	
	public long getElementAsInt64(String name) throws Exception
	{
		return this._fields.get(name).getValueAsInt64();
	}
	
	public boolean hasElement(String name)
	{
		return this._fields.containsKey(name);
	}
	
	public boolean hasElement(String name, boolean excludeNullElements)
	{
		return this._fields.containsKey(name);
	}
	
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%sfieldData = {%s", tabs, System.getProperty("line.separator")));
        
        for (Map.Entry<String, Element> item : this._fields.entrySet())
        {
        	result.append(item.getValue().prettyPrint(tabIndent + 1));
        }

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
    }
	
}

File: BEmu_cpp/src/HistoricalDataRequest/HistoricElementFieldData.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/HistoricalDataRequest/HistoricElementFieldData.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "HistoricalDataRequest/HistoricElementFieldData.h"
#include "HistoricalDataRequest/HistoricElementDateTime.h"
#include "HistoricalDataRequest/HistoricElementDouble.h"

#include "BloombergTypes/Name.h"

#include "Types/ObjectType.h"
#include "Types/IndentType.h"

namespace BEmu
{
	namespace HistoricalDataRequest
	{
		HistoricElementFieldData::HistoricElementFieldData(const Datetime& date, const std::map<std::string, ObjectType>& values)
		{
			boost::shared_ptr<HistoricElementDateTime> elmDate(new HistoricElementDateTime(date));
			boost::shared_ptr<ElementPtr> elmDateP(elmDate);

			this->_fields[elmDateP->name().string()] = elmDateP;

			for(std::map<std::string, ObjectType>::const_iterator iter = values.begin(); iter != values.end(); ++iter)
			{
				const std::string str = iter->first;
				const ObjectType obj = iter->second;
				double dvalue;

				if(obj.TryGetDouble(dvalue))
				{
					boost::shared_ptr<HistoricElementDouble> elmDouble(new HistoricElementDouble(str, dvalue));
					boost::shared_ptr<ElementPtr> elmDoubleP(elmDouble);

					this->_fields[elmDoubleP->name().string()] = elmDoubleP;
				}
			}
		}

		HistoricElementFieldData::~HistoricElementFieldData()
		{
			this->_fields.clear();
		}


		Name HistoricElementFieldData::name() const
		{
			Name result("fieldData");
			return result;
		}

		size_t HistoricElementFieldData::numValues() const
		{
			return 1;
		}

		size_t HistoricElementFieldData::numElements() const
		{
			return this->_fields.size();
		}

		SchemaElementDefinition HistoricElementFieldData::elementDefinition() const
		{
			::blpapi_DataType_t dtype = (::blpapi_DataType_t)this->datatype();
			SchemaElementDefinition result(dtype, Name("HistoricalDataRow"));
			return result;
		}

		
		bool HistoricElementFieldData::isNull() const
		{
			return false;
		}

		bool HistoricElementFieldData::isArray() const
		{
			return false;
		}

		bool HistoricElementFieldData::isComplexType() const
		{
			return true;
		}


		bool HistoricElementFieldData::hasElement(const char* name, bool excludeNullElements) const
		{
			return this->_fields.find(name) != this->_fields.end();
		}

		boost::shared_ptr<ElementPtr> HistoricElementFieldData::getElement(const char* name) const
		{
			std::string key(name);
			std::map<std::string, boost::shared_ptr<ElementPtr> >::const_iterator it = this->_fields.find(key);

			if(it == this->_fields.end())
				throw elementPtrEx;
			else
				return it->second;
		}

		int HistoricElementFieldData::getElementAsInt32(const char* name) const
		{
			return this->getElement(name)->getValueAsInt32(0);
		}

		Datetime HistoricElementFieldData::getElementAsDatetime(const char* name) const
		{
			return this->getElement(name)->getValueAsDatetime(0);
		}

		const char* HistoricElementFieldData::getElementAsString(const char* name) const
		{
			return this->getElement(name)->getValueAsString(0);
		}

		float HistoricElementFieldData::getElementAsFloat32(const char* name) const
		{
			return this->getElement(name)->getValueAsFloat32(0);
		}

		double HistoricElementFieldData::getElementAsFloat64(const char* name) const
		{
			return this->getElement(name)->getValueAsFloat64(0);
		}

		long HistoricElementFieldData::getElementAsInt64(const char* name) const
		{
			return this->getElement(name)->getValueAsInt64(0);
		}


		std::ostream& HistoricElementFieldData::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << "fieldData = {" << std::endl;
			for(std::map<std::string, boost::shared_ptr<ElementPtr> >::const_iterator iter = this->_fields.begin(); iter != this->_fields.end(); ++iter)
			{
				const boost::shared_ptr<ElementPtr> elm = iter->second;
				elm->print(stream, level + 1, spacesPerLevel);
			}

			stream << tabs << std::endl;

			return stream;
		}

	}
}